File: Leetcode/Maximal Rectangle.cc
class Solution {
public:
    static const int INF = 1<<28;
    int maximalRectangle(vector<vector<char> > &matrix)
    {
        int y = matrix.size();
        if (y == 0) {
            return 0;
        }
        int x = matrix[0].size();
        int height[y+ 5][x + 5];
        
        memset(height, 0, sizeof(height));
        
        for (int i = 0; i < y; i++) {
            for (int j = 0; j < x; j++) {
                height[i][j] = matrix[i][j] - '0';
                if (height[i][j] && i - 1 >= 0) {
                    height[i][j] += height[i-1][j];
                }
            }
        }
        
        int ans = 0;
        for (int i = 0; i < y; i++) {
            ans = max(ans, solve(height[i], x));
        }
        return ans;
    }
    
    int solve(int height[], int x)
    {
        int ans = 0;
        stack<int> st;
        height[x] = 0;
        
        for (int i = 0; i <= x; i++) {
            if (st.empty() || height[i] > height[st.top()]) {
                st.push(i);
            } else {
                while (!st.empty() && height[i] <= height[st.top()]) {
                    int t = st.top();
                    st.pop();
                    ans = max(ans, height[t] * (st.empty()? i: i - st.top() - 1));
                }
                st.push(i);
            }
        }
        return ans;
    }
};

File: Leetcode/java/Maximal Rectangle.java
public class Solution {
    public int maximalRectangle(char[][] matrix) {
        int n = matrix.length;
        if (n == 0) {
            return 0;
        }
        int m = matrix[0].length;
        if (m == 0) {
            return 0;
        }
        int graph[][] = new int[n][m];
        for (int i = 0; i < n; i++) {
            int sum = 0;
            for (int j = m - 1; j >= 0; j--) {
                if (matrix[i][j] == '1') {
                    sum++;
                } else {
                    sum = 0;
                }
                graph[i][j] = sum;
            }
        }
        int ans = 0;
        for (int i = 0; i < m; i++) {
            Stack<Integer> st = new Stack<Integer>();
            for (int j = 0; j <= n; j++) {
                int now = j == n? 0: graph[j][i];
                if (st.empty() || graph[st.peek()][i] < now) {
                    st.push(j);
                    continue;
                }
                while (!st.empty() && graph[st.peek()][i] >= now) {
                    int prev = st.pop();
                    int left = st.empty()? 0: st.peek() + 1;
                    int right = j - 1;
                    ans = Math.max(ans, graph[prev][i] * (right - left + 1));
                }
                st.push(j);
            }
        }
        return ans;
    }
}
